### ISSUE TITLE
Converter Generators Do not Emit IDs for Fields in Certain Assembly Contexts

### ISSUE THREAD

**Issue by aj-stein-nist** [usnistgov/metaschema#235]
# Describe the bug

The XSLT M4 pipeline in this repo is used by the OSCAL repository to generate converter generators themselves. For certain processing values of Metaschema `field`s in specific contexts, processing the intermediate supermodel data misses in some assembly contexts, like `/system-security-plan/system-characteristics/system-id` with its `@id` and optional `@identifier-type`, filters out and does not pass along the `@id` key to subsequent pipeline steps. This leads to an error about a malformed map where an element is missing a key, but it is not clear which key.

More detail can be found in https://github.com/usnistgov/oscal-content/pull/139#issuecomment-1242992548.

# Who is the bug affecting?

NIST OSCAL developers.

# What is affected by this bug?

Proper conversion of Metaschema-based content, specifically Metaschema, from XML to JSON.

# When does this occur?

Consistently.

# How do we replicate the issue?

1. See https://github.com/usnistgov/oscal-content/pull/139#issuecomment-1242992548 for details and explanation of this occurring as-is with CI/CD build sequencing.

# Expected behavior (i.e. solution)

The conversion of such data occurs without error, or error messaging is more explicit.

# Other Comments

N/A

**Comment by aj-stein-nist** [comment]
It was unclear to me why I had left this unassigned, but it appears I popped it off the stack of the then current sprint (Sprint 58) and left it for "the next sprint," i.e. the now current Sprint 59. I will have to circle around back to this feedback in https://github.com/usnistgov/metaschema/pull/236#issuecomment-1251128894 and make that bit happen before it is ready for final review.

**Comment by aj-stein-nist** [comment]
As agreed, @david-waltermire-nist, I know you have been back but it's been a busy week with dev and strategic planning. As discussed during planning meeting, I will circle back on this immediately upon my return on Monday 31 October 2022.

**Comment by aj-stein-nist** [comment]
Wendell and I spent a lot of last Thursday and this Monday debugging this further and I am having a lot of trouble retriggering the bug in development locally on my workstation and even in the original problem request in usnistgov/oscal-content#139, especially after rebasing its OSCAL submodule (and nested metaschema submodule). It is seeming our original hypothesis was off-target or wrong. Thankfully, taking the time to confirm the bug and a fix with tests brought us to this point. I will continue to plug away on that sample and then localize a similar construct into tests for the associated PR when I can. It is unclear how this evaporated or what I have missed, but I will update tomorrow and hopefully determine next steps then.

**Comment by aj-stein-nist** [comment]
I added a separate issue to track XProc touchups needed as determined from this work in #255.

**Comment by aj-stein-nist** [comment]
Also odd is now I keep testing and will experiment more with filters, maybe the real issue as reported "as part of usnistgov/oscal-content#139." is really usnistgov/metaschema#240 and not this issue. 🤦 

https://github.com/usnistgov/oscal-content/actions/runs/3465798740/jobs/5788953312

I need to dig deeper but now I am more unsure of myself than before.

**Comment by aj-stein-nist** [comment]
So I forgot to update this issue in the interim. I cannot seem to track this and confirm the bug with the only example in CI/CD appearing to be adjacent to #240 which breaks the CI/CD build before I can see this bug, if at all. Pulling down the artifacts show a valid JSON conversion, so I am still confused. Stepping off this and rolling back to 240 now.

/cc @wendellpiez I forgot say this explicitly here.

**Comment by aj-stein-nist** [comment]
I am going to tentatively move this up to Sprint 61, but it is unclear if this bug can be reproduced in the original context it was reported or not. I will hap to wrap this up after #240.

**Comment by aj-stein-nist** [comment]
I will come back to this and see if I can reproduce the bug that I was unable to find before.

**Comment by aj-stein-nist** [comment]
I have found some other unrelated CI/CD issues that are blocking recreating this bug and may have masked the original issue. I am continuing to investigate.

**Comment by aj-stein-nist** [comment]
I need to move this to Sprint 63 because I have to test if this is related to merging usnistgov/metaschema#240 work into OSCAL repo's submodule and bring the that repo's `develop` branch into alignment. TBD until later this week or early next week.

**Comment by aj-stein-nist** [comment]
Removing OSCAL work board per discussion with @david-waltermire-nist.

**Comment by aj-stein-nist** [comment]
I was able to pull this down again with a SSP (`oscal_leveraging-example_ssp.json`) and get back to where we were all those months back. Not sure where that leaves us with testing this issue and remediating in this repo. But is the exact scenario reported here with the JSON XDM to be converted to JSON:

```xml
<array key="system-ids"><map><string>saas_system_iaas_customer</string></map></array>
```

This is what is causing the following error when using the converter generator:

```
System ID: /home/me/code/json-xdm-to-json.xsl
Scenario: json-xdm-to-json
XSL file: /home/me/code/json-xdm-to-json.xsl
XML file: /home/me/code/gha-logs/oscal/4046515155-6959397364/git-content/src/examples/ssp/xml/oscal_leveraging-example_ssp_json-xdm.xml
Engine name: Saxon-PE 10.6
Severity: fatal
Problem ID: FOJS0006
Description: xml-to-json: Child elements of <map> must have a key attribute
Start location: 9:52
URL: http://www.w3.org/TR/2005/WD-xpath-functions-20050211/#ERRFOJS0006
```

I am using [this basic JSON in XDM form to XML converter transform](https://gist.github.com/aj-stein-nist/6914e568024242b4dde91acff132c5b1) for debugging the intermediate format.

**Comment by aj-stein-nist** [comment]
It looks like I went down the wrong rabbit hole and maybe the `super-model-to-json.xsl` stylesheet isn't the issue, certain keys for assemblies in XPath JSON form are missing. I am still investigating this but I resolved the issues with the `test-json-conversion.xpl` pipeline in incorrect/obsolete paths and figure that out more.

More to follow.

**Comment by aj-stein-nist** [comment]
> This might be happening because the `system-id` object with no flags at all would ordinarily be represented straight up, not as an object property (for the field value), so
> 
> ```
> <array key="system-ids"><string>saas_system_iaas_customer</string></array>
> ```
> 
> We need to distinguish between cases where a field casts to an object even when it has only one property (because a flag is permitted) vs when it can be 'naked' since no flags are permitted.

Weirdly our current patch in #236 fixes some of this, but there is something going on with `assembly` blocks that are converted into `group`s in the supermodel. I am currently looking into `system-information/information-types` are holding up the specific example data from the `oscal_leveraging_ssp.json` that is breaking in the oscal-content CI/CD build that is the origin of this bug report.

In short: sounds right, but there is a little more breakage. Either it moved around, or there is another part we need to fix.
